File: andriod/app/src/main/java/com/Bands70k/FirebaseSharedCommentsWrite.java
package com.Bands70k;

import android.util.Log;

import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;

import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.HashMap;
import java.util.Locale;
import java.util.Map;
import java.util.TimeZone;

/**
 * Write-only Firebase submissions for shared band descriptions.
 */
public class FirebaseSharedCommentsWrite {

    private static final String TAG = "SHARED_COMMENTS";
    private DatabaseReference mDatabase;

    public FirebaseSharedCommentsWrite() {
        mDatabase = FirebaseDatabase.getInstance().getReference();
    }

    public interface WriteCallback {
        void onComplete(boolean success);
    }

    /**
     * Writes to {@code notes/{userID}/{year}/{band}}.
     */
    public void writeSharedComment(String bandName, String descriptionText, String userName, WriteCallback callback) {
        if (staticVariables.isTestingEnv) {
            Log.d(TAG, "Skipping write in test environment");
            if (callback != null) {
                callback.onComplete(true);
            }
            return;
        }

        if (bandName == null || bandName.isEmpty()) {
            if (callback != null) {
                callback.onComplete(false);
            }
            return;
        }

        if (staticVariables.userID == null || staticVariables.userID.isEmpty()) {
            Log.e(TAG, "Missing user ID");
            if (callback != null) {
                callback.onComplete(false);
            }
            return;
        }

        if (mDatabase == null) {
            Log.e(TAG, "Firebase reference not initialized");
            if (callback != null) {
                callback.onComplete(false);
            }
            return;
        }

        int storageYear = staticVariables.resolveStorageEventYear();
        if (storageYear <= 0) {
            Log.e(TAG, "Missing eventYear in production pointer file");
            if (callback != null) {
                callback.onComplete(false);
            }
            return;
        }

        String sanitizedBandName = sanitizeBandNameForFirebase(bandName);
        String yearString = String.valueOf(storageYear);
        String path = "notes/" + staticVariables.userID + "/" + yearString + "/" + sanitizedBandName;

        SimpleDateFormat formatter = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSSXXX", Locale.US);
        formatter.setTimeZone(TimeZone.getTimeZone("UTC"));

        Map<String, Object> payload = new HashMap<>();
        payload.put("userID", staticVariables.userID);
        payload.put("userName", userName);
        payload.put("band", bandName);
        payload.put("year", yearString);
        payload.put("descriptionText", descriptionText);
        payload.put("updatedAt", formatter.format(new Date()));

        Log.d(TAG, "Writing to " + path);

        mDatabase.child(path).setValue(payload)
                .addOnSuccessListener(aVoid -> {
                    Log.d(TAG, "Write succeeded for '" + bandName + "'");
                    if (callback != null) {
                        callback.onComplete(true);
                    }
                })
                .addOnFailureListener(e -> {
                    Log.e(TAG, "Write failed: " + e.getMessage());
                    if (callback != null) {
                        callback.onComplete(false);
                    }
                });
    }

    private String sanitizeBandNameForFirebase(String bandName) {
        if (bandName == null || bandName.isEmpty()) {
            return bandName;
        }
        return bandName
                .replace(".", "_")
                .replace("#", "_")
                .replace("$", "_")
                .replace("[", "_")
                .replace("]", "_")
                .replace("/", "_")
                .replace("'", "_")
                .replace("\"", "_")
                .replace("\\", "_")
                .replaceAll("[\\p{Cntrl}]", "")
                .trim();
    }
}

File: swift/70000TonsBands/firebaseSharedCommentsWrite.swift
//
//  firebaseSharedCommentsWrite.swift
//  70K Bands
//
//  Write-only Firebase submissions for shared band descriptions.
//

import Foundation
import Firebase

class firebaseSharedCommentsWrite {

    var ref: DatabaseReference?
    private let maxAttempts = 3
    private let retryDelay: TimeInterval = 2.0

    init() {
        initializeFirebaseReference()
    }

    private func initializeFirebaseReference(attempt: Int = 1) {
        guard AppDelegate.isFirebaseConfigured else {
            print("⚠️ [SHARED_COMMENTS] Firebase not configured (attempt \(attempt)/\(maxAttempts))")
            if attempt < maxAttempts {
                DispatchQueue.global(qos: .utility).asyncAfter(deadline: .now() + retryDelay) { [weak self] in
                    self?.initializeFirebaseReference(attempt: attempt + 1)
                }
            }
            return
        }

        ref = Database.database().reference()
        print("✅ [SHARED_COMMENTS] Firebase reference initialized")
    }

    /// Writes to `notes/{userID}/{year}/{band}`.
    func writeSharedComment(
        bandName: String,
        descriptionText: String,
        userName: String,
        completion: @escaping (Bool) -> Void
    ) {
        if inTestEnvironment {
            print("⏭️ [SHARED_COMMENTS] Skipping write in test environment")
            DispatchQueue.main.async { completion(true) }
            return
        }

        guard !bandName.isEmpty else {
            DispatchQueue.main.async { completion(false) }
            return
        }

        guard let firebaseRef = ref else {
            print("❌ [SHARED_COMMENTS] Firebase reference not initialized")
            FirebaseWriteMonitor.shared.recordWriteFailure(context: "shared_comments_ref_nil")
            DispatchQueue.main.async { completion(false) }
            return
        }

        let userDataHandle = userDataHandler()
        guard !userDataHandle.uid.isEmpty else {
            print("❌ [SHARED_COMMENTS] Missing user ID")
            FirebaseWriteMonitor.shared.recordWriteFailure(context: "shared_comments_uid_empty")
            DispatchQueue.main.async { completion(false) }
            return
        }

        let sanitizedBandName = sanitizeBandNameForFirebase(bandName)
        let yearString = String(eventYear)
        let path = "notes/\(userDataHandle.uid)/\(yearString)/\(sanitizedBandName)"

        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]

        let payload: [String: Any] = [
            "userID": userDataHandle.uid,
            "userName": userName,
            "band": bandName,
            "year": yearString,
            "descriptionText": descriptionText,
            "updatedAt": formatter.string(from: Date())
        ]

        print("📝 [SHARED_COMMENTS] Writing to \(path)")

        firebaseRef.child(path).setValue(payload) { error, _ in
            if let error = error {
                print("❌ [SHARED_COMMENTS] Write failed: \(error.localizedDescription)")
                FirebaseWriteMonitor.shared.recordWriteFailure(context: "shared_comments:\(bandName)")
                DispatchQueue.main.async { completion(false) }
            } else {
                print("✅ [SHARED_COMMENTS] Write succeeded for '\(bandName)'")
                FirebaseWriteMonitor.shared.recordWriteSuccess(context: "shared_comments:\(bandName)")
                DispatchQueue.main.async { completion(true) }
            }
        }
    }

    private func sanitizeBandNameForFirebase(_ bandName: String) -> String {
        bandName
            .replacingOccurrences(of: ".", with: "_")
            .replacingOccurrences(of: "#", with: "_")
            .replacingOccurrences(of: "$", with: "_")
            .replacingOccurrences(of: "[", with: "_")
            .replacingOccurrences(of: "]", with: "_")
            .replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "'", with: "_")
            .replacingOccurrences(of: "\"", with: "_")
            .replacingOccurrences(of: "\\", with: "_")
            .components(separatedBy: .controlCharacters).joined()
            .trimmingCharacters(in: .whitespacesAndNewlines)
    }
}
